File: go/2015/day03b/day03b.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Position2D struct {
	X int32
	Y int32
}

func usage(progname string) {
	fmt.Fprintf(os.Stderr, "usage: %s <input file>\n", progname)
	os.Exit(1)
}

func process(filename string) uint32 {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	santa := Position2D{ 0, 0 };
	roboSanta := Position2D{ 0, 0 };
	positions := make(map[Position2D]struct{})
	positions[santa] = struct{}{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		i := 0
		for _, ch := range line {
			if i % 2 == 0 {
				switch ch {
				case '^':
					santa.Y += 1
				case 'v':
					santa.Y -= 1
				case '<':
					santa.X -= 1
				case '>':
					santa.X += 1
				}
				positions[santa] = struct{}{}
			} else {
				switch ch {
				case '^':
					roboSanta.Y += 1
				case 'v':
					roboSanta.Y -= 1
				case '<':
					roboSanta.X -= 1
				case '>':
					roboSanta.X += 1
				}
				positions[roboSanta] = struct{}{}
			}
			i++
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}

	return uint32(len(positions))
}

func main() {
	if len(os.Args) < 2 {
		usage(os.Args[0])
	}

	filename := os.Args[1]
	result := process(filename)
	fmt.Printf("result = %d\n", result)
}

File: pascal/2015/day03b/day03b.pas
PROGRAM Day03b(Input, Output);

USES
    Generics.Collections;

TYPE
    Position2D = RECORD
        X, Y : 	Integer
    END;

    PositionSet = SPECIALIZE THashSet<Position2D>;

VAR
    FileName : PChar;
    Res      : UInt32;

PROCEDURE Usage();
VAR
    ProgName : PChar;
BEGIN
    ProgName := argv[0];
    WriteLn('usage: ', ProgName, ' <input file>');
    Halt(1)
END;

FUNCTION Process(FileName : PChar) : UInt32;
VAR
    InFile           : Text;
    Ch               : Char;
    Santa, RoboSanta : Position2D;
    Positions        : PositionSet;
    SantaMove        : Boolean;
    Count            : UInt32;

BEGIN
    Santa.X := 0;
    Santa.Y := 0;
    RoboSanta := Santa;
    Positions := PositionSet.Create;
    Positions.Add(Santa);
    SantaMove := False;
    {$I-}
    Assign(InFile, FileName);
    Reset(InFile);
    IF IOResult <> 0 THEN
    BEGIN
        WriteLn(StdErr, 'error when opening file: ', FileName);
        Halt(1)
    END;
    {$I+}
    WHILE NOT EOF(InFile) DO
    BEGIN
        Read(InFile, Ch);
	IF SantaMove THEN
        BEGIN
            CASE Ch OF
                '^' : Santa.Y := Santa.Y + 1;
                'v' : Santa.Y := Santa.Y - 1;
                '<' : Santa.X := Santa.X - 1;
                '>' : Santa.X := Santa.X + 1;
            END;
            Positions.Add(Santa);
        END
        ELSE
        BEGIN
            CASE Ch OF
                '^' : RoboSanta.Y := RoboSanta.Y + 1;
                'v' : RoboSanta.Y := RoboSanta.Y - 1;
                '<' : RoboSanta.X := RoboSanta.X - 1;
                '>' : RoboSanta.X := RoboSanta.X + 1;
            END;
            Positions.Add(RoboSanta);
        END;
	SantaMove := NOT SantaMove;
    END;
    Close(InFile);
    Count := Positions.GetCount;
    Positions.Free;
    Process := Count
END;

BEGIN
    IF argc < 2 THEN
        Usage();
    FileName := argv[1];
    Res := Process(FileName);
    WriteLn('result = ', Res)
END.
